File: CSC330/collatz/recursiveC#/collatzR.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("No Arguments For Range Found");
            return;
        }

        int minRange, maxRange;

        // Parse command-line arguments
        if (!int.TryParse(args[0], out minRange) || !int.TryParse(args[1], out maxRange))
        {
            Console.WriteLine("Range values must be integers");
            return;
        }

        if (minRange >= maxRange)
        {
            Console.WriteLine("minRange must be less than maxRange");
            return;
        }

        var lengths = new List<int>();
        var indexLengths = new List<int>();

        // Calculate sequence lengths for each integer in the range
        for (int i = minRange; i < maxRange; i++)
        {
            int sequenceLength = CalculateLen(i);
            if (lengths.Count < 10)
            {
                lengths.Add(sequenceLength);
                indexLengths.Add(i);
            }
            else if (sequenceLength > lengths.Min())
            {
                int minIndex = lengths.IndexOf(lengths.Min());
                lengths[minIndex] = sequenceLength;
                indexLengths[minIndex] = i;
            }
        }

        // Sort by sequence length in descending order
        var sortedByLength = lengths
            .Zip(indexLengths, (length, index) => new { Length = length, Index = index })
            .OrderByDescending(x => x.Length)
            .ToList();

        Console.WriteLine("Sorted based on sequence length:");
        foreach (var item in sortedByLength)
        {
            Console.WriteLine($"{item.Index} {item.Length}");
        }

        // Sort by integer size (index) in descending order
        var sortedByIndex = lengths
            .Zip(indexLengths, (length, index) => new { Length = length, Index = index })
            .OrderByDescending(x => x.Index)
            .ToList();

        Console.WriteLine("\nSorted based on integer size:");
        foreach (var item in sortedByIndex)
        {
            Console.WriteLine($"{item.Index} {item.Length}");
        }
    }

    // Recursive function to calculate the length of the Collatz sequence for a number
    static int CalculateLen(int i)
    {
        // Base case: if i reaches 1, return the length (which is 0 because 1 is the last number)
        if (i == 1)
        {
            return 0;
        }
        // Recursive case: if i is even, divide by 2, otherwise multiply by 3 and add 1
        if (i % 2 == 0)
        {
            return 1 + CalculateLen(i / 2); // 1 step for this iteration
        }
        else
        {
            return 1 + CalculateLen(i * 3 + 1); // 1 step for this iteration
        }
    }
}

File: CSC330/collatz/recursiveJulia/collatzR.jl
function calculate_len(i::Int)
    if i == 1
        return 0
    elseif i % 2 == 0
        return 1 + calculate_len(i ÷ 2)  # Division for even numbers
    else
        return 1 + calculate_len(i * 3 + 1)  # Multiply by 3 and add 1 for odd numbers
    end
end

function main()
    # Get command-line arguments
    args = ARGS
    if length(args) < 2
        println("No Arguments For Range Found")
        return
    end
    
    min_range = 0
    max_range = 0
    # Parse command-line arguments
    try
        min_range = parse(Int, args[1])
        max_range = parse(Int, args[2])
    catch
        println("Range values must be integers")
        return
    end

    if min_range >= max_range
        println("minRange must be less than maxRange")
        return
    end

    lengths = Int[]
    index_lengths = Int[]

    # Calculate sequence lengths for each integer in the range
    for i in min_range:max_range-1
        sequence_length = calculate_len(i)
        if length(lengths) < 10
            push!(lengths, sequence_length)
            push!(index_lengths, i)
        elseif sequence_length > minimum(lengths)
            min_index = argmin(lengths)
            lengths[min_index] = sequence_length
            index_lengths[min_index] = i
        end
    end

    # Sort by sequence length in descending order
    sorted_by_length = sort(collect(zip(lengths, index_lengths)), by = x -> x[1], rev = true)

    println("Sorted based on sequence length:")
    for (length, index) in sorted_by_length
        println("$index $length")
    end

    # Sort by integer size (index) in descending order
    sorted_by_index = sort(collect(zip(lengths, index_lengths)), by = x -> x[2], rev = true)

    println("\nSorted based on integer size:")
    for (length, index) in sorted_by_index
        println("$index $length")
    end
end

# Call the main function
main()
